relay/writer/file.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package writer
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	gmysql "github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+
+	"github.com/pingcap/dm/pkg/binlog/event"
+	bw "github.com/pingcap/dm/pkg/binlog/writer"
+	"github.com/pingcap/dm/pkg/log"
+)
+
+// FileConfig is the configuration used by the FileWriter.
+type FileConfig struct {
+	RelayDir string // directory to store relay log files.
+	Filename string // the startup relay log filename, if not set then a fake RotateEvent must be the first event.
+}
+
+// FileWriter implements Writer interface.
+type FileWriter struct {
+	cfg *FileConfig
+
+	mu    sync.RWMutex
+	stage writerStage
+
+	// underlying binlog writer,
+	// it will be created/started until needed.
+	out bw.Writer
```

Comment:
use `FileWriter` directly?

Reply:
I choose to use an interface because may mock it for testing later.